lib/index.js

```diff
@@ -191,472 +156,463 @@ function Config(options) {
- *   provided context.
- */
-Config.prototype.read = function (bundleName, configName, context, callback) {
-    var self = this;
-    self._getConfigCache(bundleName, configName, context, true, function (err, config) {
-        if (err) {
-            self._getYCB(bundleName, configName, function (err, ycb) {
-                if (err) {
-                    callback(err);
+    /**
+     * Reads the contents of the named configuration file.
+     * This will auto-detect if the configuration file is YCB and read it in a context-sensitive way if so.
+     *
+     * This can possibly return the configuration object that is stored in a cache, so the caller should
+     * copy it if they intend to make a modifications.
+     * @method read
+     * @async
+     * @param {string} bundleName The bundle in which to find the configuration file.
+     * @param {string} configName Which configuration to read.
+     * @param {object} [context] The runtime context.
+     * @param {Function} callback
+     *   @param {Error|null} callback.err If an error occurred, then this parameter will
+     *   contain the error. If the operation succeeded, then `err` will be null.
+     *   @param {Object} callback.config The merged configuration object, based on the
+     *   provided context.
+     */
+    read: function (bundleName, configName, context, callback) {
+        var self = this;
+        self._getYCB(bundleName, configName, function (err, ycb) {
+            if (err) {
+                callback(err);
+                return;
+            }
+            if(self._options.baseContext) {
+                context = self._mergeBaseContext(context);
+            }
+            var key, config, groupId;
+            groupId = self._configIdMap[bundleName][configName];
+            key = bundleName + ':m:' + configName + ycb.getCacheKey(context);
+            if (self.timeAware) {
+                var now = context[self.timeDimension];
+                if (now === undefined) {
+                    callback(new Error(util.format(MESSAGES['missing time'], self.timeDimension, JSON.stringify(context))));
                     return;
                 }
+                config = self._cache.getTimeAware(key, now, groupId);
+                if(config === undefined) {
+                    config = ycb.readTimeAware(context, now, {cacheInfo: true});
+                    var expiresAt = config[self.expiresKey];
+                    if(expiresAt === undefined) {
+                        expiresAt = Number.POSITIVE_INFINITY;
+                    }
+                    if(self._options.safeMode) {
+                        config = deepFreeze(config);
+                    }
+                    self._cache.setTimeAware(key, config, now, expiresAt, groupId);
+                }
+            } else {
+                config = self._cache.get(key, groupId);
+                if(config === undefined) {
+                    config = ycb.read(context, {});
```

Comment:
i  think it was like this before, but isn't this an async execution?

Reply:
All the methods on the base ycb library are sync.